llvm/lib/Target/SPIRV/SPIRVEmitNonSemanticDI.cpp

```diff
@@ -0,0 +1,201 @@
+        FilePath = ((File->getDirectory() + sys::path::get_separator() +
+                     File->getFilename()))
+                       .str();
+        SourceLanguage = CompileUnit->getSourceLanguage();
+        break;
+      }
+    }
+    const NamedMDNode *ModuleFlags = M->getNamedMetadata("llvm.module.flags");
+    for (const auto *Op : ModuleFlags->operands()) {
+      const MDOperand &MaybeStrOp = Op->getOperand(1);
+      if (MaybeStrOp.equalsStr("Dwarf Version"))
+        DwarfVersion =
+            cast<ConstantInt>(
+                cast<ConstantAsMetadata>(Op->getOperand(2))->getValue())
+                ->getSExtValue();
+      else if (MaybeStrOp.equalsStr("Debug Info Version"))
+        DebugInfoVersion =
+            cast<ConstantInt>(
+                cast<ConstantAsMetadata>(Op->getOperand(2))->getValue())
+                ->getSExtValue();
+    }
+  }
+  // NonSemantic.Shader.DebugInfo.100 global DI instruction emitting
+  {
+    // Required LLVM variables for emitting logic
+    const SPIRVInstrInfo *TII = TM->getSubtargetImpl()->getInstrInfo();
+    const SPIRVRegisterInfo *TRI = TM->getSubtargetImpl()->getRegisterInfo();
+    const RegisterBankInfo *RBI = TM->getSubtargetImpl()->getRegBankInfo();
+    SPIRVGlobalRegistry *GR = TM->getSubtargetImpl()->getSPIRVGlobalRegistry();
+    MachineRegisterInfo &MRI = MF.getRegInfo();
+    MachineBasicBlock &MBB = *MF.begin();
+    MachineIRBuilder MIRBuilder(MBB, MBB.begin());
+
+    // Emit OpString with FilePath which is required by DebugSource
+    const Register StrReg = MRI.createVirtualRegister(&SPIRV::IDRegClass);
+    MRI.setType(StrReg, LLT::scalar(32));
+    MachineInstrBuilder MIB = MIRBuilder.buildInstr(SPIRV::OpString);
+    MIB.addDef(StrReg);
+    addStringImm(FilePath, MIB);
+
+    const SPIRVType *VoidTyMI =
+        GR->getOrCreateSPIRVType(Type::getVoidTy(*Context), MIRBuilder);
+    GR->assignSPIRVTypeToVReg(VoidTyMI, GR->getSPIRVTypeID(VoidTyMI), MF);
+
+    // Emit DebugSource which is required by DebugCompilationUnit
+    const Register DebugSourceResIdReg =
+        MRI.createVirtualRegister(&SPIRV::IDRegClass);
+    MRI.setType(DebugSourceResIdReg, LLT::scalar(32));
+    MIB = MIRBuilder.buildInstr(SPIRV::OpExtInst)
+              .addDef(DebugSourceResIdReg)
+              .addUse(GR->getSPIRVTypeID(VoidTyMI))
+              .addImm(static_cast<int64_t>(
+                  SPIRV::InstructionSet::NonSemantic_Shader_DebugInfo_100))
+              .addImm(SPIRV::NonSemanticExtInst::DebugSource)
+              .addUse(StrReg);
+    MIB.constrainAllUses(*TII, *TRI, *RBI);
+
+    const SPIRVType *I32Ty =
+        GR->getOrCreateSPIRVType(Type::getInt32Ty(*Context), MIRBuilder);
+    GR->assignSPIRVTypeToVReg(I32Ty, GR->getSPIRVTypeID(I32Ty), MF);
+
+    // Convert DwarfVersion, DebugInfo and SourceLanguage integers to OpConstant
+    // instructions required by DebugCompilationUnit
+    const Register DwarfVersionReg =
```

Comment:
Looks like worth a function that takes an integer and returns the built OpConst instruction to transform repetitive patterns into 3 function calls?

Also please add `assignSPIRVTypeToVReg()` calls.

Reply:
Discussable. In my opinion in such cases (local (not for external reuse), a lot of pointers which will be passed as function arguments, short and only three sequential uses visible all at once) introduction of additional abstraction frame isn't worth it.
Calls to `assignSPIRVTypeToVReg()` added.

Reply:
Let's then discuss please. You have three identical blocks of code, copied and pasted, except for the argument of `.addImm()`.

If you introduced an `inline` function with an integer argument it'd help to remove lines from https://github.com/llvm/llvm-project/blob/9cc733b3ac26e1dbceaa022f397fe693fe9633e9/llvm/lib/Target/SPIRV/SPIRVEmitNonSemanticDI.cpp#L148 to https://github.com/llvm/llvm-project/blob/9cc733b3ac26e1dbceaa022f397fe693fe9633e9/llvm/lib/Target/SPIRV/SPIRVEmitNonSemanticDI.cpp#L173.

Instead you'd be able to build `SPIRV::OpExtInst` (https://github.com/llvm/llvm-project/blob/9cc733b3ac26e1dbceaa022f397fe693fe9633e9/llvm/lib/Target/SPIRV/SPIRVEmitNonSemanticDI.cpp#L179) using `.addUse(my_inline_function_returns_a_register(DebugInfoVersion))` instead of `.addUse(DebugInfoVersionReg)`.

With no overheads on run-time you would be able to both reduce amount of code and avoid creating repetitive code. Please let me know your thoughts.

Reply:
We have two scenarios to consider then:
a) When newly created function is inlined.
b) When it's not inlined.

Performance (less important):
If it's inlined (a case) the resulting assembly is exactly the same.
If it's not inlined then you need to have a `call` instruction and by typical calling convention you will need to load pointers to the correct registers. Without going into much detail this kind of exchanges needs read from the stack and write to it so additional cost. Yes, you are lowering pressure on icache by reducing the amount of code generated but in this concrete scenario (three same consecutive instruction sets next to each other which have a lot of calls) it doesn't really matter because all the calls will be in icache after first OpConstantInt emission exactly because they are next to each other and the same. Plus you of course paying the at least one cycle by additional `call` and it's ceremonies.

Readability:
As I said in some of the previous posts - the tastes about what's clean and what isn't are subjective. I prefer to minimize contexts (how much I need to remember between important sections) and jumps between code pieces (maximizing locality, so I e.g. not need to have two [or more] windows open to see the full picture). Basically simplicity and pragmatism over fanciness. Introduction of additional frame raises requirement to write declaration for the new procedure and it's arguments. To ensure locality (to see all relevant code in one window) it's needed to be lambda placed above the calls. In addition to that you will need to map caller and callee arguments in so there is additional reading overhead. When we consider just three consecutive calls I think that current state is simpler approach. 

Summary: 
There are moments and places when addition of the frame is definitely worth it but this case isn't, in my opinion, one of them. 

`inline` keyword in the function context is just an information for the compiler to lower the required score for inlining the procedure. It doesn't guarantee it and it's generally deprecated (at least in the cpp community which I know) to using the keyword with such intention (it has different more specific use cases). To enforce inline a compiler specific attribute `always_inline` can be used but it will abort compilation if the function can't be inlined (we don't want that here).

Reply:
@asudarsa @michalpaszkowski I would appreciate your thoughts on this.

Reply:
I generally do agree that less code repetition is a good direction and creating a well-named function for repetitive parts is a good practice. I would not care as much about performance, the difference will be none or negligible, and `inline` does not guarantee anything -- most compilers will decide to inline or not regardless. In such isolated instance this does not make any difference to me and I am okay with either approach.

Reply:
I also agree with either approach. Those repetitive copy/pasted pieces of code may be addressed later, as a part of another PR where we add a new utility function and substitute existing copy/pasted pieces by the new function call.